src/abr/pensieve.cc

```diff
@@ -0,0 +1,105 @@
+
+      size_t chunk_size = get<1>(data_map.at(vf));
+      double chunk_size_mb = (double)chunk_size; // Bytes
+      next_chunk_sizes.push_back(make_pair(chunk_size_mb, i));
+    }
+    sort(next_chunk_sizes.begin(), next_chunk_sizes.end()); // sorts pairs by first element
+    vector<double> next_chunk_sizes_bare; // just first element of next_chunk_sizes pair
+    for (size_t i = 0; i < vformats_cnt; i++) {
+      next_chunk_sizes_bare.push_back(get<0>(next_chunk_sizes[i]));
+    }
+
+    cout << "smallest sorted chunk size: " << get<0>(next_chunk_sizes[0]) << endl;
+    cout << "largest sorted chunk size: " << get<0>(next_chunk_sizes[7]) << endl;
+    // Use this to set delivery time for last video packet
+    // Pensieve is going to consider all video packets as
+    // also containing the audio associated with those
+    // packets, which introduces some noise into these
+    // values, which perhaps we should account for, but
+    // currently do not.
+
+    cout << "Video chunk acked!!" << format << ", SSIM: "<< ssim << endl;
+    cout << "Last size per ACK: " << (double)(size) << endl;
+
+    // Okay, so when a chunk is acked we should first notify Pensieve, then wait for
+    // a response indicating what bit rate to use next
+    // RESPOND
+    // TODO: Increase trans_time to account for time to send audio chunks?
+    cout << "PBUF: " << client_.video_playback_buf() << endl;
+    json j;
+    j["delay"] = trans_time; // ms
+    j["playback_buf"] = client_.video_playback_buf(); // seconds
+    j["rebuf_time"] = client_.cum_rebuffer(); // cum seconds spent rebuffering up till now
+    j["last_chunk_size"] = (double)size; // Bytes
+    j["next_chunk_sizes"] = next_chunk_sizes_bare; //Bytes
+    uint16_t json_len = j.dump().length();
+    cout << j.dump() << endl;
+
+    connection_.write(put_field2(json_len) + j.dump());
+
+    //Now, busy wait until Pensieve responds (on my laptop this takes between 1 and 4 ms)
+    auto read_data = connection_.read_exactly(2); //read 2 byte length
+    if (read_data.empty()) {
+        cout << "Empty read, ERROR" << endl;
+    } else {
+        auto rcvd_json_len = get_uint16_2( read_data.data() );
+        cout << "Received json len: " << rcvd_json_len << endl;
+        auto read_json = connection_.read_exactly(rcvd_json_len);
+        size_t index_in_bitrate_ladder = json::parse(read_json)["bit_rate"];
+        cout << "Index of next chunk to be sent: " << index_in_bitrate_ladder << endl;
+        next_format_ = get<1>(next_chunk_sizes[index_in_bitrate_ladder]); //Works bc next_chunk_size is sorted
+        cout << "next_format: " << vformats[next_format_] << endl;
+    }
+}
+
+VideoFormat Pensieve::select_video_format()
+{
+  // Basic design here is that the next_format is set after every video ack,
+  // so any time this is called (which must come after an ack for the previous video chunk) this
+  // value will have been updated.
+  return client_.channel()->vformats()[next_format_];
```

Comment:
For now, it is fine to assume that all channels always have exactly 10 available bitrates. But notice that the channel might have changed when `select_video_format()` is called, which means that you need to record the `index_in_bitrate_ladder`, and sort the candidates of the next video chunk (of the current channel) by size.